library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,39 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer
+/// with a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must declare the same `extern` ABI
+/// string, must take the same number of arguments, and the individual argument types and the return
+/// types must be ABI-compatible.
+/// The relation of when two types are ABI-compatible is defined as follows:
+///
+/// - Every type is ABI-compatible with itself.
+/// - If `<T as Pointee>::Metadata == ()`, then `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`,
+///   `NonNull<T>` are all ABI-compatible with each other.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
```

Comment:
Here I'm being defensive just in case different ABIs might have function pointers of different size, or stuff like that...